File: backend/app/schemas/expense.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.expense import SplitType


class ExpenseSplitInput(BaseModel):
    user_id: UUID
    amount: Decimal | None = None
    percentage: Decimal | None = None
    shares: int | None = None


class ExpenseCreate(BaseModel):
    group_id: UUID

    paid_by: UUID

    title: str = Field(
        min_length=2,
        max_length=100,
    )

    description: str | None = None

    amount: Decimal = Field(gt=0)

    currency: str = Field(
        default="INR",
        max_length=10,
    )

    expense_date: datetime | None = None

    split_type: SplitType

    participants: list[ExpenseSplitInput]

    @model_validator(mode="after")
    def validate_split(self):

        # Minimum participants
        if len(self.participants) < 2:
            raise ValueError(
                "An expense must have at least two participants."
            )

        participant_ids = [
            participant.user_id
            for participant in self.participants
        ]

        # Duplicate participants
        if len(participant_ids) != len(set(participant_ids)):
            raise ValueError(
                "Duplicate participants are not allowed."
            )

        # Payer must be one of the participants
        if self.paid_by not in participant_ids:
            raise ValueError(
                "The payer must be included in participants."
            )

        # Exact split validation
        if self.split_type == SplitType.EXACT:
            if any(
                participant.amount is None
                for participant in self.participants
            ):
                raise ValueError(
                    "Amount is required for exact split."
                )

        # Percentage split validation
        if self.split_type == SplitType.PERCENTAGE:
            if any(
                participant.percentage is None
                for participant in self.participants
            ):
                raise ValueError(
                    "Percentage is required for percentage split."
                )

        # Shares split validation
        if self.split_type == SplitType.SHARES:
            if any(
                participant.shares is None
                for participant in self.participants
            ):
                raise ValueError(
                    "Shares are required for shares split."
                )

        return self


class ExpenseResponse(BaseModel):
    id: UUID
    group_id: UUID
    paid_by: UUID

    title: str
    description: str | None

    amount: Decimal

    currency: str

    split_type: SplitType

    model_config = ConfigDict(
        from_attributes=True,
    )

File: frontend/lib/features/expenses/models/expense.dart
class ExpensePayer {
  final String id;
  final String fullName;

  ExpensePayer({
    required this.id,
    required this.fullName,
  });

  factory ExpensePayer.fromJson(
    Map<String, dynamic> json,
  ) {
    return ExpensePayer(
      id: json["id"],
      fullName: json["full_name"],
    );
  }
}

class ExpenseParticipant {
  final String userId;
  final String fullName;
  final String email;
  final double amountOwed;
  final bool isSettled;

  ExpenseParticipant({
    required this.userId,
    required this.fullName,
    required this.email,
    required this.amountOwed,
    required this.isSettled,
  });

  factory ExpenseParticipant.fromJson(
    Map<String, dynamic> json,
  ) {
    return ExpenseParticipant(
      userId: json["user_id"],
      fullName: json["full_name"],
      email: json["email"],
      amountOwed: double.parse(
        json["amount_owed"].toString(),
      ),
      isSettled: json["is_settled"],
    );
  }
}

class Expense {
  final String id;
  final String groupId;

  final String title;
  final String? description;

  final double amount;
  final String currency;

  final String splitType;
  final DateTime expenseDate;

  final ExpensePayer paidBy;

  final List<ExpenseParticipant> participants;

  Expense({
    required this.id,
    required this.groupId,
    required this.title,
    required this.description,
    required this.amount,
    required this.currency,
    required this.splitType,
    required this.expenseDate,
    required this.paidBy,
    required this.participants,
  });

  factory Expense.fromJson(
    Map<String, dynamic> json,
  ) {
    return Expense(
      id: json["id"],
      groupId: json["group_id"],
      title: json["title"],
      description: json["description"],
      amount: double.parse(
        json["amount"].toString(),
      ),
      currency: json["currency"],
      splitType: json["split_type"],
      expenseDate: DateTime.parse(
        json["expense_date"],
      ),
      paidBy: ExpensePayer.fromJson(
        json["paid_by"],
      ),
      participants: (json["participants"] as List)
          .map(
            (e) => ExpenseParticipant.fromJson(e),
          )
          .toList(),
    );
  }
}
